src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,227 +9,316 @@ internal struct FileStatus
+        {
+            _initializedMainCache = -1;
+            _initializedSecondaryCache = -1;
         }
 
-        internal void SetCreationTime(string path, DateTimeOffset time)
+        internal bool IsDirectory(ReadOnlySpan<char> path, bool continueOnError = false)
         {
-            // Unix provides APIs to update the last access time (atime) and last modification time (mtime).
-            // There is no API to update the CreationTime.
-            // Some platforms (e.g. Linux) don't store a creation time. On those platforms, the creation time
-            // is synthesized as the oldest of last status change time (ctime) and last modification time (mtime).
-            // We update the LastWriteTime (mtime).
-            // This triggers a metadata change for FileSystemWatcher NotifyFilters.CreationTime.
-            // Updating the mtime, causes the ctime to be set to 'now'. So, on platforms that don't store a
-            // CreationTime, GetCreationTime will return the value that was previously set (when that value
-            // wasn't in the future).
-            SetLastWriteTime(path, time);
+            // We first check if main path has directory flag, then follow the symbolic link in case the target has directory flag
+            // So we need to try to refresh both caches
+            EnsureCachesInitialized(path, continueOnError);
+            return _isDirectory; // Value should be set in Refresh
         }
 
-        internal DateTimeOffset GetLastAccessTime(ReadOnlySpan<char> path, bool continueOnError = false)
+        internal bool IsHidden(ReadOnlySpan<char> path, bool continueOnError = false)
         {
-            EnsureStatInitialized(path, continueOnError);
-            if (!_exists)
-                return DateTimeOffset.FromFileTime(0);
-            return UnixTimeToDateTimeOffset(_fileStatus.ATime, _fileStatus.ATimeNsec);
+            // We only check for the hidden flag in the main path without following symbolic links
+            if (!IsMainCacheValid)
+            {
+                if (!TryRefreshMainCache(path))
+                {
+                    if (!continueOnError)
+                    {
+                        ThrowOnCacheInitializationError(path);
+                    }
+                }
+            }
+            return HasHiddenFlag;
         }
 
-        internal void SetLastAccessTime(string path, DateTimeOffset time) => SetAccessOrWriteTime(path, time, isAccessTime: true);
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            // We only check for the readonly flag in the main path without following symbolic links
+            if (!IsMainCacheValid)
+            {
+                if (!TryRefreshMainCache(path))
+                {
+                    if (!continueOnError)
+                    {
+                        ThrowOnCacheInitializationError(path);
+                    }
+                }
+            }
```

Comment:
Can we use one `if`? Or move all the logic in `TryRefreshMainCache()`?

Reply:
I can move it all to one `if`. I don't think we should throw from within a try method.

Reply:
I mean `IsMainCacheValid` check could be in `TryRefreshMainCache()`.